File: src/main/resources/web/js/zk/video/Video.js
zk.video.Video = zk.$extends(zk.Widget, {

    _src: '',
    _muted: false,
    _autoplay: false,
    _controls: false,
    _loop: false,
    _preload: 'none',
    _poster: '',
    _crossorigin: '',
    _playbackRate: 1,
    _currentTime: '',
    getSrc: function () {
        return this._src;
    },
    setSrc: function (value) {
        if (this._src != value) {
            this._src = value;
        }
    },
    getCurrentTime: function () {
        return this._currentTime;
    },
    setCurrentTime: function (value) {
        if (this._currentTime != value) {
            this._currentTime = value;
        }
    },
    getPreload: function () {
        return this._preload;
    },
    setPreload: function (value) {
        if (this._preload != value) {
            this._preload = value;
        }
    },
    getPoster: function () {
        return this._poster;
    },
    getPlaybackRate: function () {
        return this._playbackRate;
    },
    setPlaybackRate: function (value) {
        try {
            var ctx = $('#'+this.uuid + '-video');
            if (ctx) {
                ctx.playbackRate = value;
            }
        } catch (b) {
            zk.error(b.message || b)
        }
    },
    setPoster: function (value) {
        if (this._poster != value) {
            this._poster = value;
        }
    },
    getCrossorigin: function () {
        return this._crossorigin;
    },
    setCrossorigin: function (value) {
        if (this._crossorigin != value) {
            this._crossorigin = value;
        }
    },
    getMuted: function () {
        return this._muted;
    },
    setMuted: function (value) {
        if (this._muted != value) {
            this._muted = value;
        }
    },
    getAutoplay: function () {
        return this._autoplay;
    },
    setAutoplay: function (value) {
        if (this._autoplay != value) {
            this._autoplay = value;
        }
    },
    getControls: function () {
        return this._controls;
    },
    setControls: function (value) {
        if (this._controls != value) {
            this._controls = value;
        }
    },
    getLoop: function () {
        return this._loop;
    },
    setLoop: function (value) {
        if (this._loop != value) {
            this._loop = value;
        }
    },
    setPlaying: function (value) {
        var ctx = $('#'+this.uuid + '-video');
        if (ctx) {
       //     ctx.currentTime = value;
        }
    },
    setVolume: function (value) {
        var ctx = $('#'+this.uuid + '-video');
        if (ctx) {
            ctx.volume = value;
        }
    },
    setMuted: function (value) {
        var ctx = $('#'+this.uuid + '-video');
        if (ctx) {
            ctx.muted = value;
        }
    },
    setCurrentTime: function (value) {
        var ctx = $('#'+this.uuid + '-video');
        if (ctx) {
            ctx.currentTime = parseFloat(value);
        }
    },
    bind_: function () {
        this.$supers('bind_', arguments);
        var elementId = '#'+this.uuid + '-video';
        var ctx = $('#'+this.uuid + '-video');

            ctx.attr('src', this._src);
            this.setPlaybackRate(this._playbackRate);
            ctx.on('play', function () {
                var currentTime = this.currentTime;
                var n = '#' + this.id + '-video';
                var widget = zk.Widget.$(n);
                zAu.send(new zk.Event(widget, 'setPlaying', {playing: 'true', currentTime: currentTime}, {toServer: true}));
            });
            ctx.on('pause', function () {
                var currentTime = this.currentTime;
                var n = '#' + this.id + '-video';
                var widget = zk.Widget.$(n);
                zAu.send(new zk.Event(widget, 'setPlaying', {playing: 'false', currentTime: currentTime}, {toServer: true}));
            });

    },
    unbind_: function () {
        this.$supers('unbind_', arguments);
    },
    domAttrs_: function () {
        var a = '';
        if(this._muted)
            a += ' muted ';
        if(this._autoplay )
            a += ' autoplay ';
        if(this._controls)
            a += ' controls ';
        if(this._loop)
            a += ' loop ';
        if(this._playsinline)
            a += ' playsinline ';
        if(this._preload)
            a += ' preload=\"' + this._preload + '\" ';
        if(this._poster)
            a+=' poster=\"' + this._poster + '\" ';
        if(this._crossorigin)
            a += ' crossorigin=\"' + this._crossorigin + '\" ';
        if(this._style)
            a += ' style=\"' + this._style + '\" ';
        if(this._height)
            a += ' height=\"' + this._height + '\" ';

        a += ' class=\"' + this._sclass + '\" '
        return a
    }
});

File: src/main/java/com/gitlab/pedrioko/core/zk/component/Video.java
package com.gitlab.pedrioko.core.zk.component;

import lombok.Data;
import org.zkoss.zk.au.out.AuInvoke;
import org.zkoss.zk.ui.HtmlBasedComponent;
import org.zkoss.zk.ui.sys.ContentRenderer;

import java.io.IOException;

public @Data
class Video extends HtmlBasedComponent {
    private String src = "";
    private boolean controls = false;
    private boolean muted = false;
    private boolean autoplay = false;
    private boolean loop = false;
    private String preload = "";
    private String poster = "";
    private String crossorigin = "";
    private double playbackRate = 1.0D;


    public Video() {
    }

    public Video(String source) {
        this.src = source;
    }

    @Override
    protected void renderProperties(ContentRenderer renderer) throws IOException {
        super.renderProperties(renderer);
        render(renderer, "src", src);
        render(renderer, "controls", controls);
        render(renderer, "muted", muted);
        render(renderer, "autoplay", autoplay);
        render(renderer, "loop", loop);
        render(renderer, "preload", preload);
        render(renderer, "poster", poster);
        render(renderer, "crossorigin", crossorigin);
        if (this.playbackRate != 1.0D) {
            this.render(renderer, "playbackRate", this.playbackRate);
        }}

    public void setSrc(String src) {
        this.src = src;
        smartUpdate("src", src);

    }

    public void setMute(boolean muted) {
        this.response((String)null, new AuInvoke(this, "setMuted", muted));
    }

    public void setPlaying(boolean playing) {
        this.response((String) null, new AuInvoke(this, "setPlaying", playing));
    }

    public void setCrossorigin(String crossorigin) {
        this.crossorigin = "use-credentials".equalsIgnoreCase(crossorigin) ? "use-credentials" : "anonymous";
        if (!this.crossorigin.equals(crossorigin)) {
            this.crossorigin = crossorigin;
            this.smartUpdate("crossorigin", (Object) this.crossorigin);
        }
    }

    public void setPreload(String preload) {
        preload = "none".equalsIgnoreCase(preload) ? "none" : ("metadata".equalsIgnoreCase(preload) ? "metadata" : "auto");
        if (!preload.equals(this.preload)) {
            this.preload = preload;
            this.smartUpdate("preload", (Object) this.preload);
        }
    }

    public void setPoster(String poster) {
        if (poster != null && !poster.equals(this.poster)) {
            this.poster = poster;
            this.smartUpdate("poster", (Object) this.poster);
        }
    }

    public void setAutoplay(boolean autoplay) {
        if (this.autoplay != autoplay) {
            this.autoplay = autoplay;
            this.smartUpdate("autoplay", this.autoplay);
        }
    }

    public void setControls(boolean controls) {
        if (this.controls != controls) {
            this.controls = controls;
            this.smartUpdate("controls", this.controls);
        }
    }

    public void setLoop(boolean loop) {
        if (this.loop != loop) {
            this.loop = loop;
            this.smartUpdate("loop", this.loop);
        }
    }

    public void setMuted(boolean muted) {
        if (this.muted != muted) {
            this.muted = muted;
            this.smartUpdate("muted", this.muted);
        }
    }

    public void setPlaybackRate(double playbackRate) {
        if (this.playbackRate != playbackRate) {
            this.playbackRate = playbackRate;
            this.smartUpdate("playbackRate", this.playbackRate);
        }
    }
}
